bionic/code_references.py

```
@@ -0,0 +1,124 @@
+                else:
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_DEREF", "LOAD_CLOSURE"]:
+                set_tos(context.cells[op.argval])
+            elif op.opname == "IMPORT_NAME":
+                try:
+                    set_tos(importlib.import_module(op.argval))
+                except ImportError:
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR", "IMPORT_FROM"]:
+                if tos is None:
+                    refs.append(op.argval)
+                elif isinstance(tos, str):
+                    tos += "." + op.argval
+                else:
+                    tos = getattr(tos, op.argval)
+            elif op.opname == "DELETE_FAST" and tos:
+                del context.varnames[op.argval]
+                tos = None
+            elif op.opname == "STORE_FAST" and tos:
+                context.varnames[op.argval] = tos
+                tos = None
+            elif op.opname == "LOAD_FAST" and op.argval in context.varnames:
+                set_tos(context.varnames[op.argval])
+            else:
+                # For all other instructions, hash the current TOS.
+                if tos is not None:
+                    refs.append(tos)
+                    tos = None
+        except Exception as e:
+            message = oneline(
+                f"""
+            Bionic found a code reference in file ${code.co_filename}
+            at line ${lineno} that it cannot hash when hashing
+            ${code.co_name}. We will ignore hashing the code reference,
+            which means any changes to the reference won't invalidate
+            the cache.
+
+            This is most likely a bug in Bionic. Please raise a new
+            issue at https://github.com/square/bionic/issues to let us
+            know.
+            """
+            )
+            message = message + "\n" + str(e)
+            warnings.warn(message)
```

I wasn't sure if I should throw an error or warning here, but I ended up using a warning. I think it's really bad if a user can't use the flow at all. Maybe we should give more information like how to delete / invalidate the cache entries manually when they see this warning. That way, the users won't be completely blocked. Wdyt? 

FWIW, Streamlit throws an error here.

---

It sounds like we're going to keep iterating through the remaining opcodes. Does that make sense? Will our state still be valid?

---

Now that I had some more time to think about it, I think I should change the catch block and add the tos to references if it contains one.

---

After thinking about it a bit more (and talking on slack), I changed it to an error (like Streamlit).